pkg/proxy/iptables/proxier_test.go

```diff
@@ -4473,10 +4472,14 @@ COMMIT
 :KUBE-NODEPORTS - [0:0]
 :KUBE-POSTROUTING - [0:0]
 :KUBE-MARK-MASQ - [0:0]
+:KUBE-SVL-AQI2S6QIMU7PVVRP - [0:0]
 -A KUBE-POSTROUTING -m mark ! --mark 0x4000/0x4000 -j RETURN
 -A KUBE-POSTROUTING -j MARK --xor-mark 0x4000
 -A KUBE-POSTROUTING -m comment --comment "kubernetes service traffic requiring SNAT" -j MASQUERADE
 -A KUBE-MARK-MASQ -j MARK --or-mark 0x4000
+-A KUBE-SERVICES -m comment --comment "ns1/svc1 cluster IP" -m tcp -p tcp -d 172.30.1.1 --dport 80 -j KUBE-SVL-AQI2S6QIMU7PVVRP
+-A KUBE-SVL-AQI2S6QIMU7PVVRP -m comment --comment "ns1/svc1 cluster IP" -m tcp -p tcp -d 172.30.1.1 --dport 80 ! -s 10.0.0.0/8 -j KUBE-MARK-MASQ
+-A KUBE-SVL-AQI2S6QIMU7PVVRP -m comment --comment "ns1/svc1 has no local endpoints" -j KUBE-MARK-DROP
```

Comment:
The change to drop rather than reject will have some negative consequences - clients have to time out.  I swear we used this in a test, but I can't find where, so maybe it went away.  Anyone think this will be a problem?  @aojea ?

Reply:
It will have consequences, but I was assuming they were positive...

Why is there an internal client trying to connect to a local endpoint that doesn't exist? There are basically two scenarios: (a) the client is confused and should not be trying to connect to the service from this node; (b) the client is correct, and there _should_ be an endpoint on this node, but for some reason there isn't. In the first case, rejecting the connection is the best answer, but it seems like the second case is more likely (eg, the endpoint pod is restarting) and in that case, dropping is better because it will let the client retry and eventually succeed.

(Note: eTP:Local also drops rather than rejecting, so this makes internal and external traffic policy behave the same. Also, I'm pretty sure there was no explicit decision to make iTP:Local do reject rather than drop before; it just happened that way because of how iTP was implemented.)

Reply:
> dropping is better because it will let the client retry and eventually succeed.

Why wouldn't a client retry in case of REJECT?  I guess they'd get a different error code...

Reply:
I think that silently dropping makes the retry to depend on the OS timers, and is not the best solution ... the client code should handle these situations and we should inform the client something is wrong

Also, the REJECT sends a RST that clear conntrack entries all the way down ... 


The e2e test is here https://github.com/kubernetes/kubernetes/blob/515f8342d342ce5d017df26838d6dd36fb162361/test/e2e/network/service.go#L1788

Reply:
And in fact, [the internal traffic policy KEP](https://github.com/kubernetes/enhancements/tree/master/keps/sig-network/2086-service-internal-traffic-policy#proposal) actually says that traffic is supposed to be dropped:

> Introduce a new field in Service `spec.internalTrafficPolicy`. The field will have 2 codified values:
> 
> 1. Cluster (default): route to all cluster-wide endpoints (or use topology aware subsetting if enabled).
> 2. Local: only route to node-local endpoints, drop otherwise.

Reply:
I don't think we've alweays been precise about "drop" vs "reject".  Yeah, the OS will retry EVENTUALLY.  But so will clients.

Anyway, we can do a followup discussion here.

Reply:
I understand Dan reasoning, but I'm not sure those assumptions that this will only happen on race conditions are always true
> under the assumption that we hit a race condition (endpoint exited but LB hasn't re-polled node health again yet) and if we let the client try again then it will (eventually) succeed

I think that is better to notify clients because it encourage client applications to handle network errors, also, defaults OS values are very conservative

https://blog.cloudflare.com/when-tcp-sockets-refuse-to-die/

> The retries are staggered at 1s, 3s, 7s, 15s, 31s, 63s marks (the inter-retry time starts at 2s and then doubles each time). By default the whole process takes 130 seconds, until the kernel gives up with the ETIMEDOUT errno

dropping for me is something you use for firewalling, when you don't care at all about the clients ... but this was always a controversial topic :smile: